File: Dynamic Programming/583. Delete Operation for Two Strings.py
"""
583. Delete Operation for Two Strings

"""

class Solution:
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        n, m = len(word1), len(word2)
        dp = [[0,]*(n+1) for _ in range(m+1)]
        for i in range(n):
            dp[0][i+1] = i+1;
        for i in range(m):
            dp[i+1][0] = i+1;    
        for i in range(m):
            for j in range(n):
                dp[i+1][j+1] = dp[i][j] if word1[j] == word2[i] else min(dp[i][j+1], dp[i+1][j])+1
        #print(dp)
        return dp[-1][-1]

class Solution:
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        n, m = len(word1), len(word2)
        dp = [0,]*(n+1)
        for i in range(n):
            dp[i+1] = i+1;
        for i in range(m):
            dp[0] = i+1
            prev = i
            for j in range(n):
                prev, dp[j+1] = dp[j+1], prev if word1[j] == word2[i] else min(dp[j], dp[j+1])+1
        #print(dp)
        return dp[-1]
        
# Bottom-up
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        m, n = len(word1), len(word2)
        dp = [[0] * (n + 1) for i in range(m + 1)]
        for i in range(m):
            for j in range(n):
                dp[i + 1][j + 1] = max(dp[i][j + 1], dp[i + 1][j], dp[i][j] + (word1[i] == word2[j]))
        return m + n - 2 * dp[m][n]

# Top-Down
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        memo = {}
        def topDown(i, j):
            if (i, j) not in memo:
                if i == 0 or j == 0:
                    ans = i if i else j 
                elif word1[i-1] == word2[j-1]:
                    ans = topDown(i-1, j-1)
                else:
                    ans = 1 + min(topDown(i-1, j), topDown(i, j-1))
                memo[(i, j)] = ans
            return memo[(i, j)]
        return topDown(len(word1), len(word2))

# Top-Down
# i == len(A) or j == len(B), one of the strings is empty, so the answer is just the sum of the remaining lengths.
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        memo = {}
        def topDown(i, j):
            if (i, j) not in memo:
                if i == len(word1) or j == len(word2):
                    ans = len(word1) + len(word2) - i - j
                elif word1[i] == word2[j]:
                    ans = topDown(i+1, j+1)
                else:
                    ans = 1 + min(topDown(i+1, j), topDown(i, j+1))
                memo[(i, j)] = ans
            return memo[(i, j)]
        return topDown(0, 0)

File: Dynamic Programming/583. Delete Operation for Two Strings.cpp
/*
         s1 s e a 
     s2  0  1 2 3 
     e   1  2 1 2
     a   2  3 2 1  
     t   3  4 3 2
 
    如果s[i]!=s[j]  dp[i][j] = 1+min(dp[i-1][j], dp[i][j-1]) =  dp[i][j-1] + 1 横向移动，代表删除的是s1
    如果s[i]!=s[j]  dp[i][j] = 1+min(dp[i-1][j], dp[i][j-1]) =  dp[i-1][j] + 1 纵向移动，代表删除的是s2

*/

class Solution {
public:
    int minDistance(string word1, string word2) {
        int n1 = word1.size(), n2 = word2.size();
        vector<vector<int>>dp(n1+1, vector<int>(n2+1));
        for(int i = 0; i<=n1; ++i)
            dp[i][0] = i;
        for(int j = 0; j<=n2; ++j)
            dp[0][j] = j;
        
        for(int i = 1; i<=n1; ++i){
            for(int j = 1; j<=n2; ++j){
                if(word1[i-1] == word2[j-1])
                    dp[i][j] = dp[i-1][j-1];
                else
                    dp[i][j] = min(dp[i-1][j], dp[i][j-1])+1;
            }
        }
        return dp[n1][n2];
    }
};

class Solution {
public:
    int minDistance(string word1, string word2) {
        int n1 = word1.size(), n2 = word2.size();
        vector<vector<int>>dp(2, vector<int>(n2+1,INT_MAX));
        for(int i = 0; i<=n1; i++)
            for(int j = 0, k = i%2; j<=n2; j++)
                dp[k][j] =  i == 0 || j ==0 ? (i == 0? j: i) : word1[i-1] == word2[j-1] ? dp[k^1][j-1]: 1+min(dp[k^1][j], dp[k][j-1]); 
        return dp[n1%2][n2];
    }
};


class Solution {
public:
    int minDistance(string word1, string word2) {
        vector<vector<int>>dp(word1.size()+1, vector<int>(word2.size()+1,-1));
        return topDown(dp, word1,word2, word1.size(), word2.size());
    }
    
    int topDown(vector<vector<int>>&dp, const string& word1, const string& word2, int i , int j){
        if(dp[i][j] >= 0)
            return dp[i][j];
        if(i == 0 || j== 0)
            return i == 0 ? j: i;
        if(word1[i-1] == word2[j-1])
            dp[i][j] = topDown(dp, word1, word2, i-1, j-1);
        else
            dp[i][j] = 1 + min(topDown(dp, word1, word2, i-1,j), topDown(dp, word1, word2, i, j-1));
           
        return dp[i][j];
    }
};

//TopDown
//i == len(A) or j == len(B), one of the strings is empty, so the answer is just the sum of the remaining lengths.
class Solution {
public:
    int minDistance(string word1, string word2) {
        vector<vector<int>>dp(word1.size()+1, vector<int>(word2.size()+1,-1));
        return topDown(dp, word1,word2, 0, 0);
    }
    
    int topDown(vector<vector<int>>&dp, const string& word1, const string& word2, int i , int j){
        if(dp[i][j] >= 0)
            return dp[i][j];
        if(i == word1.size() || j== word2.size())
            return word1.size() + word2.size() -i - j;
        if(word1[i] == word2[j])
            dp[i][j] = topDown(dp, word1, word2, i+1, j+1);
        else
            dp[i][j] = 1 + min(topDown(dp, word1, word2, i+1,j), topDown(dp, word1, word2, i, j+1));
        return dp[i][j];
    }
};


//(Longest Common Subsequence)
/*
To make them identical, just find the longest common subsequence. 
The rest of the characters have to be deleted from the both the strings, 
    which does not belong to longest common subsequence.

 */

class Solution {
public:
    int minDistance(string word1, string word2) {
        vector<vector<int>>dp(word1.size()+1, vector<int>(word2.size()+1));
        for(int i = 1; i <= word1.size(); i++)
            for(int j = 1; j <= word2.size(); j++) 
                dp[i][j] = word1[i-1] == word2[j-1] ? dp[i-1][j-1] + 1 : max(dp[i-1][j], dp[i][j-1]);
        return word1.size() + word2.size() - 2 * dp[word1.size()][word2.size()];
    }
};